src/main/haskell/kore/src/Kore/Builtin/Map.hs

```diff
@@ -349,3 +376,133 @@ lookupSymbolInKeys
+empty, otherwise this equation is just a trivial shuffling and does not actually
+make progress toward simplification. We introduce special cases when @x₁@ and/or
+@x₂@ is missing.
+ -}
+-- TODO (thomas.tuegel): Handle the case of two framed maps.
+unify
+    :: forall level variable m p expanded proof.
+       ( Eq (variable Object), Show (variable Object)
+       , SortedVariable variable
+       , MonadCounter m
+       , MetaOrObject level
+       , p ~ PureMLPattern level variable
+       , expanded ~ ExpandedPattern level variable
+       , proof ~ SimplificationProof level
+       )
+    => MetadataTools level StepperAttributes
+    -> (p -> p -> Result (m (expanded, proof)))
+    -> (p -> p -> Result (m (expanded, proof)))
+unify
+    tools@MetadataTools { symbolOrAliasSorts }
+    simplifyChild
+  =
+    unify0
+  where
+    bottom :: (expanded, proof)
+    bottom = (ExpandedPattern.bottom, SimplificationProof)
+
+    hookTools = StepperAttributes.hook <$> tools
+
+    unifyWith unifyChild as bs =
+        do
+            let as' = Map.difference as bs
+                bs' = Map.difference bs as
+            cs <- sequence (Map.intersectionWith unifyChild as bs)
+            return (as', bs', cs)
+
+    unify0
+        (DV_ sort' (BuiltinDomainMap map1))
+        (DV_ _     (BuiltinDomainMap map2))
+      =
+        unifyConcrete sort' map1 map2
+    unify0
+        (DV_ sort' (BuiltinDomainMap map1))
+        (App_ concat'
+            [ DV_ _ (BuiltinDomainMap map2)
+            , x@(Var_ _)
+            ]
+        )
+        | isSymbolConcat hookTools concat' =
+            unifyFramed1 sort' map1 concat' map2 x
+    unify0 app_@(App_ _ _) dv_@(DV_ _ _) = unify0 dv_ app_
+    unify0 _ _ = empty
+
+    unifyConcrete
+        :: ( level ~ Object
+           , k ~ ConcretePurePattern Object
+           )
+        => Sort level
+        -> Map k p
+        -> Map k p
+        -> Result (m (expanded, proof))
+    unifyConcrete sort' map1 map2 = do
+        (rem1, rem2, _quot) <- unifyWith simplifyChild map1 map2
+        let
+            unified = give symbolOrAliasSorts $ do
+                _quot <- Map.map fst <$> sequence _quot
+                let q = asBuiltinDomainValue sort' <$> sequenceA _quot
+                return (q, SimplificationProof)
+        return
+            (if Map.null rem1 && Map.null rem2
+                then unified
+                else return bottom
+            )
+
+    unifyFramed1
+        :: ( level ~ Object
+           , k ~ ConcretePurePattern Object
+           )
+        => Sort level
+        -> Map k p
+        -> SymbolOrAlias level
+        -> Map k p
+        -> p
+        -> Result (m (expanded, proof))
+    unifyFramed1 sort' map1 concat' map2 x = do
+        do
+            (rem1, rem2, _quot) <- unifyWith simplifyChild map1 map2
+            -- Unify the elements missing from map2 with the framing variable
+            _rem <- simplifyChild x (asBuiltinDomainValue sort' rem1)
```

Comment:
We did discuss this. The keys of a Map or Set are functional. The values of a Map or List are not (necessarily) functional. When I made the keys concrete, I made changes to `Kore.Step.PatternAttributes` so that the check for functionality will descend into the Map and List values.

Reply:
ok, thanks.